### PR TITLE
CD-1921 update sdk

### PR DESCRIPTION
**Pull request by yepchaos**

### CODE REVIEW

File: src/sdk.ts [hunk 1]
```diff
@@ -267,20 +271,36 @@ export class CyanSDK {
...
-     * Approve Payment Plan Contract to spend a specific ERC721 token on behalf of the signer.
-     * @param {string} address - The address of the ERC721 contract.
+     * Approve Payment Plan Contract to spend a specific ERC721 or ERC1155 token on behalf of the signer.
+     * @param {string} address - The address of the ERC721 or ERC1155 contract.
      * @param {string} tokenId - The ID of the token to approve.
      * @returns {Promise<void>} - A Promise that resolves when the transaction is confirmed.
      * @throws {Error} - Throws an error if the PaymentPlan contract is not found or if the transaction fails.
      */
-    public async getApproval(address: string, tokenId: string): Promise<void> {
+    public async getApproval(address: string, tokenId: string, tokenType?: INftType): Promise<void> {
```
Comment (Qlio) [inline src/sdk.ts]: change to:`tokenType: INftType = INftType.ERC721`

File: src/sdk.ts [hunk 2]
```diff
@@ -267,20 +271,36 @@ export class CyanSDK {
...
+        if (tokenType == INftType.ERC1155) {
+            const erc1155Contract = ERC1155Factory.connect(address, this.signer);
+            const isApprovedForAll = await erc1155Contract.isApprovedForAll(
+                this.signer.getAddress(),
+                paymentPlanContract.address
+            );
+
+            if (isApprovedForAll) return;
+            tx = await erc1155Contract.setApprovalForAll(paymentPlanContract.address, true);
+        } else {
```
Comment (Qlio) [inline src/sdk.ts]: check is erc721 and if not erc1155 or erc721, throw error: **`Not supported token type ${tokenType}`**

File: src/sdk.ts [hunk 3]
```diff
@@ -267,20 +271,36 @@ export class CyanSDK {
...
-
-        if (approvedTo.toLowerCase() === paymentPlanContract.address.toLowerCase()) return;
-
-        const tx = await erc721Contract.approve(paymentPlanContract.address, tokenId);
+        let tx: ContractTransaction | undefined = undefined;
+        if (tokenType == INftType.ERC1155) {
+            const erc1155Contract = ERC1155Factory.connect(address, this.signer);
+            const isApprovedForAll = await erc1155Contract.isApprovedForAll(
+                this.signer.getAddress(),
+                paymentPlanContract.address
```
Comment (Qlio) [inline src/sdk.ts]: shouldn't it be a conduit address?

File: src/sdk.ts [hunk 4]
```diff
@@ -267,20 +271,36 @@ export class CyanSDK {
...
+            );
+
+            if (isApprovedForAll) return;
+            tx = await erc1155Contract.setApprovalForAll(paymentPlanContract.address, true);
+        } else {
+            const erc721Contract = ERC721Factory.connect(address, this.signer);
+            const isApprovedForAll = await erc721Contract.isApprovedForAll(
+                this.signer.getAddress(),
+                paymentPlanContract.address
```
Comment (Qlio) [inline src/sdk.ts]: ditto.

**Review by Qlio — CHANGES_REQUESTED** [review]

File: src/sdk.ts [hunk 5]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+            })
+        );
+        const totalRepaymentAmountByCurrency = plansWithNextPayment.reduce<{
+            [key: string]: {
+                totalCost: BigNumber;
+                currency: ICurrency;
+                isNativeCurrency: boolean;
+            };
+        }>((acc, cur) => {
+            if (acc[cur.currency.address.toLowerCase()]) {
```
Comment (Qlio) [inline src/sdk.ts]: create constant `const currencyAddress = cur.currency.address.toLowerCase()`;

File: src/sdk.ts [hunk 6]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+                };
+            })
+        );
+        const totalRepaymentAmountByCurrency = plansWithNextPayment.reduce<{
+            [key: string]: {
+                totalCost: BigNumber;
+                currency: ICurrency;
+                isNativeCurrency: boolean;
+            };
+        }>((acc, cur) => {
```
Comment (Qlio) [inline src/sdk.ts]: change `cur` to `plan`

File: src/sdk.ts [hunk 7]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+            if (acc[cur.currency.address.toLowerCase()]) {
+                acc[cur.currency.address.toLowerCase()] = {
+                    totalCost: acc[cur.currency.address.toLowerCase()].totalCost.add(
+                        cur.nextPayment?.currentPayment || 0
+                    ),
+                    currency: cur.currency,
+                    isNativeCurrency: cur.currency.address.toLowerCase() === ethers.constants.AddressZero.toLowerCase(),
```
Comment (Qlio) [inline src/sdk.ts]: no need for `toLowerCase` because zero address is not case sensitive.

File: src/sdk.ts [hunk 8]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+                    bulkTransactions.push({
+                        to: currency.address,
+                        data: encodedAllowanceFnData,
+                        value: 0,
+                    });
+                }
+            }
+        }
+        for (const plan of plansWithNextPayment) {
+            const isNativeCurrency = plan.currency.address.toLowerCase() === ethers.constants.AddressZero.toLowerCase();
```
Comment (Qlio) [inline src/sdk.ts:539]: no need for lowercase.

File: src/sdk.ts [hunk 9]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+
+        const currencyValues = Object.values(totalRepaymentAmountByCurrency);
+        const cyanWallet = await this.getCyanWallet();
+        const contract = PaymentPlanV2Factory.connect(plans[0].paymentPlanContractAddress, this.signer);
+
+        const bulkTransactions = [];
+        for (const { currency, totalCost, isNativeCurrency } of currencyValues) {
+            if (!isNativeCurrency) {
+                const erc20 = ERC20Factory.connect(currency.address, this.signer);
+                const userAllowance = await erc20.allowance(this.signer.getAddress(), cyanWallet.address);
```
Comment (Qlio) [inline src/sdk.ts]: why should the user approve the ERC20 token to cyanWallet?

File: src/sdk.ts [hunk 10]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+                if (userAllowance.lt(totalCost)) {
+                    const encodedAllowanceFnData = erc20.interface.encodeFunctionData('approve', [
```
Comment (Qlio) [inline src/sdk.ts:526]: this will not work. Please check by paying non native currency item.

File: src/sdk.ts [hunk 11]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+                    currency: cur.currency,
+                    isNativeCurrency: cur.currency.address.toLowerCase() === ethers.constants.AddressZero.toLowerCase(),
+                };
+            }
+            return acc;
+        }, {});
+
+        const currencyValues = Object.values(totalRepaymentAmountByCurrency);
+        const cyanWallet = await this.getCyanWallet();
+        const contract = PaymentPlanV2Factory.connect(plans[0].paymentPlanContractAddress, this.signer);
```
Comment (Qlio) [inline src/sdk.ts]: check all plan's paymentPlanContractAddress. throw error if one of them has a different paymentPlanContractAddress.

File: src/sdk.ts [hunk 12]
```diff
@@ -407,6 +435,152 @@ export class CyanSDK {
...
+                releaseWallet &&
+                cyanWallet.address.toLowerCase() !== releaseWallet.toLowerCase()
+            ) {
+                if (plan.tokenType === ItemType.ERC1155) {
+                    const contractIFace = ERC1155Factory.createInterface();
+                    const encodedFnDataTransfer = contractIFace.encodeFunctionData('safeTransferFrom', [
+                        cyanWallet.address,
+                        releaseWallet,
+                        plan.tokenId,
+                        1,
```
Comment (Qlio) [inline src/sdk.ts]: plan should have token amount field.

File: src/sdk.ts [hunk 13]
```diff
@@ -267,20 +271,44 @@ export class CyanSDK {
...
-
-        if (approvedTo.toLowerCase() === paymentPlanContract.address.toLowerCase()) return;
-
-        const tx = await erc721Contract.approve(paymentPlanContract.address, tokenId);
+        let tx: ContractTransaction | undefined = undefined;
+        const cyanConduitAddress = this.configs.cyanConduitAddresses.filter(
+            cyanConduit => cyanConduit.chainId === chainId
+        );
+        const approvalAddress =
+            cyanConduitAddress.length > 0 ? cyanConduitAddress[0].address : paymentPlanContract.address;
```
Comment (Qlio) [inline src/sdk.ts]: let's not give permission to paymentPlanContract. just use conduit address.

File: src/sdk.ts [hunk 14]
```diff
@@ -267,20 +271,44 @@ export class CyanSDK {
...
+            const balanceOf = await erc1155Contract.balanceOf(this.signer.getAddress(), tokenId);
+            if (balanceOf.eq(0)) return;
+
+            const isApprovedForAll = await erc1155Contract.isApprovedForAll(this.signer.getAddress(), approvalAddress);
+            if (isApprovedForAll) return;
+            tx = await erc1155Contract.setApprovalForAll(approvalAddress, true);
+        } else if (tokenType === ItemType.ERC721) {
+            const erc721Contract = ERC721Factory.connect(address, this.signer);
+            const ownerOf = await erc721Contract.ownerOf(tokenId);
+            if (ownerOf.toLowerCase() === cyanWallet.address.toLowerCase()) return;
```
Comment (Qlio) [inline src/sdk.ts:296]: why cyan wallet?

File: src/sdk.ts [hunk 15]
```diff
@@ -267,20 +271,44 @@ export class CyanSDK {
...
+            const isApprovedForAll = await erc721Contract.isApprovedForAll(this.signer.getAddress(), approvalAddress);
+            if (isApprovedForAll) return;
+            const approvedTo = await erc721Contract.getApproved(tokenId);
+
+            if (approvedTo.toLowerCase() === approvalAddress.toLowerCase()) return;
+            tx = await erc721Contract.approve(approvalAddress, tokenId);
```
Comment (Qlio) [inline src/sdk.ts]: we should give set approval for all

File: src/sdk.ts [hunk 16]
```diff
@@ -267,20 +271,44 @@ export class CyanSDK {
...
+        const chainId = await this.signer.getChainId();
         const paymentPlanContract = await this.getPaymentPlanContract();
-        const erc721Contract = ERC721Factory.connect(address, this.signer);
-        const approvedTo = await erc721Contract.getApproved(tokenId);
-
-        if (approvedTo.toLowerCase() === paymentPlanContract.address.toLowerCase()) return;
-
-        const tx = await erc721Contract.approve(paymentPlanContract.address, tokenId);
+        let tx: ContractTransaction | undefined = undefined;
+        const cyanConduitAddress = this.configs.cyanConduitAddresses.filter(
```
Comment (Qlio) [inline src/sdk.ts]: return map(chainId => conduitAddress) from config.

**Review by Qlio — CHANGES_REQUESTED** [review]

File: src/sdk.ts (hunk 14 above)
Comment (yepchaos) [reply src/sdk.ts:296]: If that token is in cyan wallet, we don't have to give approval.

File: src/sdk.ts (hunk 16 above)
Comment (yepchaos) [reply src/sdk.ts]: I used the 'find', just like the other configs.

**Pull request merged**